File: medium/0981_time_based_key_value_store/solution.cpp
#include <algorithm>
#include <string>
#include <unordered_map>
#include <vector>

class TimeMap {
public:
  TimeMap() {
    timestampsMap = std::unordered_map<std::string, std::vector<int>>();
    valuesMap = std::unordered_map<int, std::string>();
  }

  void set(std::string key, std::string value, int timestamp) {
    timestampsMap[key].push_back(timestamp);
    valuesMap[timestamp] = value;
  }

  std::string get(std::string key, int timestamp) {
    if (timestampsMap.find(key) == timestampsMap.end())
      return "";

    auto it = std::lower_bound(timestampsMap[key].begin(),
                               timestampsMap[key].end(), timestamp);

    if (it != timestampsMap[key].end() && *it <= timestamp)
      return valuesMap[*it];
    else if (it == timestampsMap[key].end() ||
             (it != timestampsMap[key].begin() && *it > timestamp))
      return valuesMap[*(it - 1)];
    else
      return "";
  }

private:
  std::unordered_map<std::string, std::vector<int>> timestampsMap;
  std::unordered_map<int, std::string> valuesMap;
};

File: medium/0981_time_based_key_value_store/solution.py
from bisect import bisect_left
from collections import defaultdict


class TimeMap:
    def __init__(self):
        self.timestampsMap = defaultdict(list)
        self.valuesMap = {}

    def set(self, key: str, value: str, timestamp: int) -> None:
        self.timestampsMap[key].append(timestamp)
        self.valuesMap[timestamp] = value

    def get(self, key: str, timestamp: int) -> str:
        if key not in self.timestampsMap:
            return ""

        i = bisect_left(self.timestampsMap[key], timestamp)

        if (
            i < len(self.timestampsMap[key])
            and self.timestampsMap[key][i] <= timestamp
        ):
            return self.valuesMap[self.timestampsMap[key][i]]
        elif i > 0 and self.timestampsMap[key][i - 1] <= timestamp:
            return self.valuesMap[self.timestampsMap[key][i - 1]]

        return ""
